### PR TITLE
Copy some read tests to tests/functional

### PR DESCRIPTION
**Pull request by serhiikobalt**
Online parts of ApiTest & CatalogClientTest copied to functional tests

Relates-to: OLPEDGE-719
Signed-off-by: Serhii Lysenko <ext-serhii.lysenko@here.com>

### CODE REVIEW

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+  ASSERT_LT(0, dataResponse.GetResult()->size());
+  std::string dataStr(dataResponse.GetResult()->begin(),
+                      dataResponse.GetResult()->end());
+  ASSERT_EQ("DT_2_0031", dataStr);
+}
+
+TEST_P(CatalogClientTest, GetDataWithPartitionIdVersion2) {
+  olp::client::HRN hrn(GetTestCatalog());
+
+  auto catalogClient =
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]:
```suggestion
  auto catalog_client =
```

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+      std::make_unique<olp::dataservice::read::CatalogClient>(hrn, settings_);
+
+  auto request = olp::dataservice::read::DataRequest();
+  request.WithLayerId("testlayer").WithPartitionId("269").WithVersion(2);
+  auto dataResponse =
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]:
```suggestion
  auto data_response =
```

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+  ASSERT_LT(0, dataResponse.GetResult()->size());
+  std::string dataStr(dataResponse.GetResult()->begin(),
+                      dataResponse.GetResult()->end());
+  ASSERT_EQ("DT_2_0031", dataStr);
+}
+
+TEST_P(CatalogClientTest, GetDataWithPartitionIdInvalidVersion) {
+  olp::client::HRN hrn(GetTestCatalog());
+
+  auto catalogClient =
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]:
```suggestion
  auto catalog_client =
```

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 4]
```diff
@@ -0,0 +1,598 @@
...
+        return future.GetFuture().get();
+      });
+
+  ASSERT_FALSE(partitionsResponse.IsSuccessful());
+  ASSERT_EQ(olp::client::ErrorCode::BadRequest,
+            partitionsResponse.GetError().GetErrorCode());
+  ASSERT_EQ(400, partitionsResponse.GetError().GetHttpStatusCode());
+
+  request.WithVersion(-1);
+  partitionsResponse =
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]:
```suggestion
  partitions_response =
```

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp [hunk 5]
```diff
@@ -0,0 +1,455 @@
...
+  std::promise<olp::dataservice::read::ApiClientLookup::ApiClientResponse>
+      configClientPromise;
+  olp::dataservice::read::ApiClientLookup::LookupApiClient(
+      olp::client::OlpClientFactory::Create(*settings_), "config", "v1", hrn,
+      [&configClientPromise](
+          olp::dataservice::read::ApiClientLookup::ApiClientResponse response) {
+        configClientPromise.set_value(response);
+      });
+
+  auto clientResponse = configClientPromise.get_future().get();
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]:
```suggestion
  auto client_response = configClientPromise.get_future().get();
```

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp [hunk 6]
```diff
@@ -0,0 +1,455 @@
...
+      };
+
+  auto startTime = std::chrono::high_resolution_clock::now();
+  auto cancelFn = olp::dataservice::read::ConfigApi::GetCatalog(
+      configClient, GetTestCatalog(), boost::none, catalogCallback);
+  auto catalogResponse = catalogPromise.get_future().get();
+  auto end = std::chrono::high_resolution_clock::now();
+
+  std::chrono::duration<double> time = end - startTime;
+  std::cout << "duration: " << time.count() * 1000000 << " us" << std::endl;
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: I would replace std::cout with trace

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+ * License-Filename: LICENSE
+ */
+
+#include <chrono>
+#include <iostream>
+#include <regex>
+#include <string>
+#include <tuple>
+
+#include <gmock/gmock.h>
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]:
```suggestion
```
Do we need mock here?

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+#include <olp/core/client/OlpClientFactory.h>
+#include <olp/core/client/OlpClientSettingsFactory.h>
+#include <olp/core/http/Network.h>
+#include <olp/core/http/NetworkRequest.h>
+#include <olp/core/http/NetworkResponse.h>
+#include <olp/core/logging/Log.h>
+#include <olp/core/porting/make_unique.h>
+#include <olp/core/utils/Dir.h>
+#include <olp/dataservice/read/PrefetchTilesRequest.h>
+#include <testutils/CustomParameters.hpp>
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]: In ApiTest.cpp you moved this include  (CustomParameters.hpp) in a separate group. So, better to stick with that approach

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+    return resultStream.str();
+  }
+
+  template <typename T>
+  T GetExecutionTime(std::function<T()> func) {
+    auto startTime = std::chrono::high_resolution_clock::now();
+    auto result = func();
+    auto end = std::chrono::high_resolution_clock::now();
+    std::chrono::duration<double> time = end - startTime;
+    std::cout << "duration: " << time.count() * 1000000 << " us" << std::endl;
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]: Do we need to log the duration of function execution? I suggest to remove it in order to not clutter the log.

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+      });
+
+  ASSERT_TRUE(dataResponseCompressed.IsSuccessful())
+      << PrintError(dataResponseCompressed.GetError());
+  ASSERT_LT(0u, dataResponseCompressed.GetResult()->size());
+  ASSERT_EQ(dataResponse.GetResult()->size(),
+            dataResponseCompressed.GetResult()->size());
+}
+
+void dumpTileKey(const olp::geo::TileKey& tileKey) {
```
Comment (borzun) [inline tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]: Move to anonymous namespace.
```suggestion
void DumpTileKey(const olp::geo::TileKey& tileKey) {
```

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp (hunk 9 above)
Comment (serhiikobalt) [reply tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]: I'd prefer to not change the test's behaviour in this PR. This part was implemented intentionally, at least requires more discussion before cutting it off.

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp [hunk 11]
```diff
@@ -0,0 +1,455 @@
...
+
+#include "ApiClientLookup.h"
+#include "generated/api/BlobApi.h"
+#include "generated/api/ConfigApi.h"
+#include "generated/api/MetadataApi.h"
+#include "generated/api/PlatformApi.h"
+#include "generated/api/QueryApi.h"
+#include "generated/api/VolatileBlobApi.h"
+
+class ApiTest : public ::testing::Test {
```
[the remark this answers is not in the document]
Comment (borzun) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: Why?

[the remark this answers is not in the document]
Comment (serhiikobalt) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: reduced link time

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp (hunk 6 above)
Comment (serhiikobalt) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: I recall we decided to not use sdk logs in tests

File: tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp (hunk 9 above)
Comment (borzun) [reply tests/functional/olp-cpp-sdk-dataservice-read/CatalogClientTest.cpp]: What about usage of `OLP_SDK_LOG_TRACE` in this case? The `std::cout` won't produce anything on the Android platform.

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp (hunk 6 above)
Comment (borzun) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: For some platforms the `std::cout` won't produce anything

File: tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (diachenko-mischa) [reply tests/functional/olp-cpp-sdk-dataservice-read/ApiTest.cpp]: @borzun, it will save us from equal symbols with different implementations.

**Pull request merged**